system/jlib/jtrace.cpp

```diff
@@ -107,6 +108,260 @@ class NoopSpanExporterFactory
     }
 };
 
+/**
+ * Converts an OpenTelemetry span kind to its string representation.
+ * These are OTel span kinds defined in api/include/opentelemetry/trace/span.h,
+ * not HPCC JLib CSpan kinds
+ *
+ * @param spanKind The OpenTelemetry span kind to convert.
+ * @return The string representation of the OpenTelemetry span kind.
+ */
+const char * spanKindToString(opentelemetry::trace::SpanKind spanKind)
+{
+    switch (spanKind)
+    {
+        case opentelemetry::trace::SpanKind::kClient:
+            return "Client";
+        case opentelemetry::trace::SpanKind::kServer:
+            return "Server";
+        case opentelemetry::trace::SpanKind::kProducer:
+            return "Producer";
+        case opentelemetry::trace::SpanKind::kConsumer:
+            return "Consumer";
+        case opentelemetry::trace::SpanKind::kInternal:
+            return "Internal";
+        default:
+            return "Unknown";
+    }
+}
+
+class JLogSpanExporter final : public opentelemetry::sdk::trace::SpanExporter
+{
+private:
+    SpanLogFlags logFlags = SpanLogFlags::LogNone;
+
+public:
+    JLogSpanExporter(SpanLogFlags spanLogFlags) : shutDown(false)
+    {
+        logFlags = spanLogFlags;
+    }
+
+    /**
+    * @return Returns a unique pointer to an empty recordable object
+    */
+    virtual std::unique_ptr<opentelemetry::sdk::trace::Recordable> MakeRecordable() noexcept override
+    {
+        return std::unique_ptr<opentelemetry::sdk::trace::Recordable>(new opentelemetry::sdk::trace::SpanData());
+    }
+
+    /**
+    * Export - Formats recordable spans in HPCC Jlog format and reports to JLog
+    *
+    * @param recordables
+    * @return Always returns success
+    */
+    opentelemetry::sdk::common::ExportResult Export(
+      const nostd::span<std::unique_ptr<opentelemetry::sdk::trace::Recordable>> &recordables) noexcept override
+    {
+        if (isShutDown())
+        {
+            ERRLOG("JLog Trace Exporter: Export failed, exporter is shutdown");
+            return opentelemetry::sdk::common::ExportResult::kFailure;
+        }
+
+        for (auto &recordable : recordables)
+        {
+            auto span = std::unique_ptr<opentelemetry::sdk::trace::SpanData>(
```

Comment:
Great question, I naively relied on the approach accepted by most of the established OTel exporter implementations:

```
https://github.com/open-telemetry/opentelemetry-cpp/blob/25343e6e3ba5e85645684bc36d18001eddfc9700/exporters/otlp/src/otlp_recordable_utils.cc#L67C89-L67C96
https://github.com/open-telemetry/opentelemetry-cpp/blob/064fef0d871c57ffac6739d3311659a5770a9db4/exporters/ostream/src/span_exporter.cc#L59
https://github.com/open-telemetry/opentelemetry-cpp/blob/25343e6e3ba5e85645684bc36d18001eddfc9700/exporters/elasticsearch/src/es_log_record_exporter.cc#L345
https://github.com/open-telemetry/opentelemetry-cpp/blob/25343e6e3ba5e85645684bc36d18001eddfc9700/exporters/zipkin/src/zipkin_exporter.cc#L62C32-L62C42
```

Looking a little deeper, the opentelemetry::sdk::trace::Recordable seems safe to static_cast:

https://github.com/open-telemetry/opentelemetry-cpp/blob/25343e6e3ba5e85645684bc36d18001eddfc9700/sdk/include/opentelemetry/sdk/trace/recordable.h#L168

Can you expand on this comment/question, and perhaps provide exactly what you think should be done to ensure cast safety and std::unique_ptr.release operation? Thanks.

Reply:
I don't know the open telemetry api, but if it passes and array of recordable rather than span it implies to me that there could be items in that list that are not spans.  If that is not true, then it is worth including a comment to that effect in the code.

Reply:
I believe the array is populated by the MakeRecordable method implemented by the exporter (L170)